Fix simultaneous usage of clearcoat and tangent data

Oletus opened the pull request:
Clearcoat normals need to use temporary variable names that are different from regular normals so that there's no name conflict. This also moves sampling of the normal map outside of the normal perturbation helper function so that there isn't as much duplicate code.

Oletus committed:
Fix simultaneous usage of clearcoat and tangent data

Clearcoat normals need to use temporary variable names that are different from regular normals so that there's no name conflict. This also moves sampling of the normal map outside of the normal perturbation helper function so that there isn't as much duplicate code.

Oletus committed:
Share tangent space matrix between regular and clearcoat normals

In case a material has both regular and clearcoat normal maps and is using vertex tangents, we can compute the tangent space matrix made out of the three basis vectors once and use it for both regular and clearcoat normals.

Oletus commented:
Not sure what's up with the CI, looks like it maybe timed out due to some unrelated condition?

Mugen87 commented:
Current (https://threejs.org/examples/webgl_materials_physical_clearcoat):

![image](https://user-images.githubusercontent.com/12612165/67789116-b49ecd80-fa73-11e9-9b04-b2780e1861dd.png)

PR:

![image](https://user-images.githubusercontent.com/12612165/67789152-c41e1680-fa73-11e9-988b-22099b2690be.png)

Oletus committed:
Always use the correct normal map for clearcoat

Prior to this PR clearcoat used "normalMap" in case per-vertex tangents were enabled, and "clearcoatNormalMap" in case per-vertex tangents were disabled. Now we're correctly always using "clearcoatNormalMap".

Oletus commented:
Looks like the previous code also had another bug - in case per-vertex tangents were being used, clearcoat would use the regular normal map. I used that as a reference so I accidentally broke the case that was previously working correctly. Added a commit to always use the right normal map now.

WestLangley commented:
>It seems the actual normalMap is now visible.

@Mugen87 No, that is a bug in the PR.

WestLangley commented:
@Oletus said
>Clearcoat normals need to use temporary variable names that are different from regular normals so that there's no name conflict.

Can you please point to the bug you were referring to -- and presumably fixed?

//

I spent about an hour studying your PR and I agree your refactoring is an improvement. Thanks!

Oletus commented:
I didn't file a separate bug report for the bugs.

Bug 1: there's a shader name conflict when `vertexTangents` are enabled on a material with clearcoat. You can reproduce this by using `BufferGeometryUtils.computeTangents` and enabling vertexTangents in the clearcoat example.

Bug 2: clearcoat shader code under the `USE_TANGENT` define was also referring to regular normal map instead of the clearcoat normal map , this line in the old code: 

`vec3` mapN = texture2D( normalMap, vUv ).xyz * 2.0 - 1.0;`

WestLangley commented:
@Oletus This looks good to me. Thank you.

It is not your fault, but IMO, the shader code with the chunks is very fragile. Logical errors and edge cases are easy to overlook. I found the bug you introduced prior to you mentioning it (bug 2), but I did not notice the redeclaration bug you fixed (bug 1) until I asked.

I think it would be beneficial if @arobertson0 would review these changes, too.

arobertson0 commented:
The examples are running correctly, and the code is clear and seemingly correct =)

I'm noticing that the contents of `clearcoat_normal_fragment_maps` and `normal_fragment_maps` are very similar.
Would it be worthwhile to abstract the logic as a function? It would probably land in `normal_fragment_begin`, right where _vTBN_ is being defined.

WestLangley commented:
>The examples are running correctly

The examples worked before...

The related examples don't use tangents. Testing requires a lot of effort, and numerous use cases.

Oletus commented:
Should we add something to examples to have at least a basic test for this? It would be easy enough to add variations of geometry that use computed per-vertex tangents to the webgl_materials_physical_clearcoat example.

WestLangley commented:
>Should we add something to examples

That is up to @mrdoob.

For the record, there are 8 cases:
- with/without normal map
- with/without clearcoat normal map
- with/without tangents

Mugen87 commented:
I don't think the mentioned examples should block this PR from being merged. They can also be added at a later point, too.

mrdoob commented:
Looks good to me. We can improve things as we go.

mrdoob commented:
Thanks!

WestLangley commented:
/ping @sunag FYI.  :-)
